include/seqan3/range/views/pseudo_random_access.hpp

```diff
@@ -0,0 +1,199 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::views::pseudo_random_access.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/range/views/detail.hpp>
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\interface seqan3::detail::pseudo_random_access_range <>
+ * \extends   std::ranges::range
+ * \brief     This concept checks if a type models a random access range via `begin_ra` and `end_ra` member
+ *            functions, albeit constant access time might not be guaranteed.
+ * \ingroup   range
+ *
+ * \details
+ *
+ * This concept describes the requirements a range must fulfil in order to be used as a random access range.
+ *
+ * ### Concepts and doxygen
+ *
+ * The requirements for this concept are given as related functions and type traits.
+ * Types that model this concept are shown as "implementing this interface".
+ */
+/*!\name Requirements for seqan3::detail::pseudo_random_access_range
+ * \brief You can expect these functions on all types that model seqan3::pseudo_random_access_range.
+ * \relates seqan3::detail::pseudo_random_access_range
+ * \{
+ */
+/*!\fn pseudo_random_access_range::begin_ra()
+ * \brief Returns a pseudo random access iterator.
+ *
+ * \details
+ *
+ * The returned iterator must model the std::random_access_iterator, although accessing elements through this iterator
+ * might not be constant.
+ *
+ * \attention This is a concept requirement, not an actual function (however types
+ *            modelling this concept will provide an implementation).
+ */
+/*!\fn pseudo_random_access_range::end_ra()
+ * \brief Returns a pseudo random access sentinel.
+ *
+ * \details
+ *
+ * The returned sentinel must model std::sentinel_for with the corresponding iterator obtained by begin_ra().
+ *
+ * \attention This is a concept requirement, not an actual function (however types
+ *            modelling this concept will provide an implementation).
+ */
+//!\cond
+template <typename rng_t>
+SEQAN3_CONCEPT pseudo_random_access_range = std::ranges::range<rng_t> && requires (rng_t rng)
+{
+    rng.begin_ra();
+    rng.end_ra();
+
+    requires std::random_access_iterator<decltype(rng.begin_ra())>;
+    requires std::sentinel_for<decltype(rng.begin_ra()), decltype(rng.end_ra())>;
```

Comment:
shouldn't this be sized_sentinal since I want it to be distractable from the begin iterator?

Reply:
No, that is a little tricky. We only want to know if the returned iterator is an random_access_iterator (so you can compute the distance between two iterators of the same type) and that the returned sentinel is a sentinel for this iterator. That does not necessarily mean, that `sentinel - iterator` is also valid. Simply put, there could be a random access range that is not sized, e.g. if it is an infinite range like repeat. So size'dness of a range is orthogonal to whether or not it is random access.

Reply:
Phew, true :+1: